tsl/src/hypercore/hypercore_handler.c

```diff
@@ -3353,6 +3355,77 @@ hypercore_scan_sample_next_tuple(TableScanDesc scan, SampleScanState *scanstate,
 	return false;
 }
 
+/*
+ * Cleanup callback for Hypercore conversion state.
+ *
+ * The cleanup happens when the conversion state's memory context is
+ * destroyed. It ensures cleanup of tuplesort state, unless it was already
+ * performed.
+ */
+static void
+conversionstate_cleanup(void *arg)
+{
+	ConversionState *state = arg;
+
+	if (state->tuplesortstate)
+	{
+		tuplesort_end(state->tuplesortstate);
+		state->tuplesortstate = NULL;
+	}
+
+	if (conversionstate)
+	{
+		Assert(state == conversionstate);
+		conversionstate = NULL;
+	}
+}
+
+static ConversionState *
+conversionstate_create(const HypercoreInfo *hcinfo, const Relation rel)
+{
+	CompressionSettings *settings = ts_compression_settings_get(hcinfo->compressed_relid);
+	Tuplesortstate *tuplesortstate;
+	MemoryContext mcxt;
+	MemoryContext oldmcxt;
+	ConversionState *state;
+
+	oldmcxt = MemoryContextSwitchTo(PortalContext);
+	/*
+	 * We want to ensure the tuplesort state is cleaned up by calling
+	 * tuplesort_end() in case of failures. This is necessary to release disk
+	 * resources.
+	 *
+	 * A memory context callback is used for this purpose. The callback is
+	 * attached to the Hypercore conversion state. The tuplesort state will
+	 * allocate its own child context, but they cannot be children of the
+	 * conversion memory context because children are freed before the
+	 * parent. Instead, make both the tuplesort and conversion state children
+	 * of PortalContext. Since they are destroyed in reverse order, the memory
+	 * context callback for the conversion state can call tuplesort_end()
+	 * before the tuplesort is freed.
```

Comment:
I tried it, and it is possible. But I ran into some strange issues. I can check again.

Reply:
So, the reason why making it a child is that it creates a circular memory release problem.

Basically, when  `tuplesort_end()` is called due to normal processing, it will destroy its memory context, including child contexts first. But if the child then also calls `tuplesort_end()` to clean up the parent it doesn't work.

Reply:
Ah, IC.